### ISSUE TITLE
The tool occasionally corrupts files being copied

### ISSUE THREAD

**Issue by msuhanov** [Velocidex/velociraptor#2153]
Hello.

When collecting data in real-world environments using Velociraptor, I observed the following corruption scenarios.

1. Version: 0.6.5

_One case:_ the $MFT file is empty (0 bytes). Other files are collected properly.
_Another case:_ the $MFT file does not start with expected data (many file record segments located in the beginning of the source $MFT file are not present in its copy – the first kibibytes of the file are absent).
_Two more cases:_ the Amcache hive has missing hive bins and existing (successfully copied) hive bins become misaligned somewhere in the middle of the copied file.

The .idx file for one case with the Amcache hive is:
```
{"file_offset":0,"original_offset":0,"file_length":131072,"length":131072}
{"file_offset":131072,"original_offset":131072,"file_length":163840,"length":163840}
{"file_offset":229376,"original_offset":294912,"file_length":229376,"length":229376}
{"file_offset":458752,"original_offset":524288,"file_length":262144,"length":262144}
{"file_offset":720896,"original_offset":786432,"file_length":262144,"length":262144}
{"file_offset":983040,"original_offset":1048576,"file_length":131072,"length":131072}
{"file_offset":1114112,"original_offset":1179648,"file_length":0,"length":131072}
```

As you can see, `file_offset` is not equal to `original_offset` after the first two runs.

(Another .idx file has similar misaligned runs.)

2. Version: 0.6.6

I was able to reproduce this (or a very similar) bug using the debugger while slowing down the execution flow of the Velociraptor binary. In particular, the following condition is occasionally met (before the corresponding data run is fully read):

https://github.com/Velocidex/velociraptor/blob/999b62de13332c9e7afcced09d54c8808ebabfbd/utils/copy.go#L70

In the previous version (0.6.5), the `maybeCollectSparseFile` routine (`reporting/container.go`) does not add padding bytes for partial reads, thus creating misaligned data. In the current version (0.6.6), there is no misaligned data, but there are null bytes instead of real data, because of the padding:

https://github.com/Velocidex/velociraptor/blob/999b62de13332c9e7afcced09d54c8808ebabfbd/reporting/container.go#L327

![Example](https://user-images.githubusercontent.com/7497142/195160317-9255c7d8-b2c5-40bd-b526-0b0af79f664d.png)

(If I run the same binary again, on the same system, the error can disappear or affect a different file.)

---

The nature of this bug is unknown to me. It affects existing data collection procedures.

**Comment by scudette** [comment]
This is expected as sparse files mean that there is a gap in the file. Velociraptor does not pad the gap instead it copies the runs that contain data and create an idx file that tells it which runs are present. The file offset refers to the offset in the collected file while the original offset refers to the offset in the original file. If they are not equal this is because there is a sparse run in the original that was skipped. NTFS has many such sparse runs all the time so this happens in many cases. Examples I have seen include mft and registry files and occasionally evtx files.

In previous versions downloading the file from the uploads tab in the GUI will pad out the gaps while the export zip file will include the idx file.

In the next version users have a choice to also pad out files in the export download zip.

**Comment by scudette** [comment]
It looks like a read from the underlying NTFS parser is returning less data than we expect - so we pad it to maintain alignment. This might be an ntfs parser bug or just a side effect of the system changing while the acquisition is taking place (for example in the ntuser.dat file the MFT might have the runs changed between the time we parse the mft and try to copy the file out).

Are you saying that repeated acquisitions sometimes work better?

**Comment by msuhanov** [comment]
> This is expected as sparse files mean that there is a gap in the file.

These gaps and missing data are not caused by sparse ranges.

> If they are not equal this is because there is a sparse run in the original that was skipped. NTFS has many such sparse runs all the time so this happens in many cases. Examples I have seen include mft and registry files and occasionally evtx files.

There are no sparse ranges in corresponding gaps or at locations with missing non-zero data.

---

For example, the Amcache file mentioned in the bug report has the following .idx file:

```
{"file_offset":0,"original_offset":0,"file_length":131072,"length":131072}
{"file_offset":131072,"original_offset":131072,"file_length":163840,"length":163840}
{"file_offset":229376,"original_offset":294912,"file_length":229376,"length":229376}
{"file_offset":458752,"original_offset":524288,"file_length":262144,"length":262144}
{"file_offset":720896,"original_offset":786432,"file_length":262144,"length":262144}
{"file_offset":983040,"original_offset":1048576,"file_length":131072,"length":131072}
{"file_offset":1114112,"original_offset":1179648,"file_length":0,"length":131072}
```

The corresponding mapping pairs (data runs) are:
- as raw bytes (taken from the $MFT file using a HEX editor): `31 20 DF AF 7A 31 28 1E 78 DD 31 38 57 F3 31 31 40 DB D6 92 31 40 8C AC FE 31 40 93 C9 FE 00`;
- as decoded (offset, size) tuples (in clusters): `[(8040415, 32), (5777405, 40), (9050964, 56), (1897007, 64), (1810107, 64), (1730638, 64)]`;
- as decoded (offset, size) tuples (in bytes): `[(32933539840, 131072), (23664250880, 163840), (37072748544, 229376), (7770140672, 262144), (7414198272, 262144), (7088693248, 262144)]`.

As you can see, there are no sparse ranges, every offset points to a real chain of clusters. (However, the Velociraptor tool reports the last range as sparse, because the file size is larger than the valid data size. The code treats this situation as a sparse range, here: `https://github.com/Velocidex/go-ntfs/blob/f791e9dd76f7725f8f3a1a55da2e0700cbf5b80e/parser/attribute.go#L78`. But this range does not cause any problems during the copy operation.)

Again, this is not caused by sparse ranges, because some non-zero data is missing in the copy.

---

> It looks like a read from the underlying NTFS parser is returning less data than we expect - so we pad it to maintain alignment.

Yes, this is a bug.

> This might be an ntfs parser bug or just a side effect of the system changing while the acquisition is taking place (for example in the ntuser.dat file the MFT might have the runs changed between the time we parse the mft and try to copy the file out).

The underlying mapping pairs are intact. The `Read` routine just returns zero in some situations (the exact line is mentioned in the original report).

> Are you saying that repeated acquisitions sometimes work better?

Yes.

**Comment by scudette** [comment]
The read returns less data than we requested. I am not sure why this would be yet but there is nothing else to do than pad the runs with zeros.

**Comment by msuhanov** [comment]
It suddenly returns zero bytes in all cases where this bug is triggered (no other situations have been observed in the debugger).

> I am not sure why this would be yet but there is nothing else to do than pad the runs with zeros.

My best guess is that there is an issue with threads that results in zero bytes being read while there is more actual data to be read. (Like suddenly closing the handle to the source drive. But this is just a guess, given the "floating" nature of the bug: e.g., I can execute the binary 5 times and get correct results, then execute it 5 more times and get 3 collections with errors of this kind.)

**Comment by scudette** [comment]
This is actually not an unlikely analysis - the file handle to the actual device is retained in a cache which is refreshed periodically (to ensure that we read fresh data all the time - this is a live system afterall and we dont want to deal with cache incoherency). 

It is possible that there is a read ongoing during when the file handle is refreshed. Let me try to trigger this in a test.

Here is the closer loop:
https://github.com/Velocidex/velociraptor/blob/707f3edb3b1d6b513b63902895f0086f7b881e17/accessors/ntfs/readers/ntfs_reader.go#L75

You can make the lifetime very long by setting in the VQL
```sql
LET NTFS_CACHE_TIME <= 1000000
```
Will essentially disable this feature.

How long does the acquisition take before you tend to hit these errors? The default lifetime is 60 seconds.

**Comment by scudette** [comment]
Yes I just verified that when setting NTFS_CACHE_TIME to 1 second I can clearly cause this failure to occur every time

![Windows_11_Test_VM_2022_2](https://user-images.githubusercontent.com/3856546/195330224-d8d9bfa9-3230-4d7e-92b6-a5acbb9d6283.png)

**Comment by msuhanov** [comment]
> How long does the acquisition take before you tend to hit these errors? The default lifetime is 60 seconds.

No more than two minutes in most cases.

**Comment by scudette** [comment]
In my testing simply retrying the read operation will work because the ntfs context will be forced to rebuild, resent caches etc. In any case I also think that there is no need to reset the context every 60 seconds for a KapeFiles collection because it is mostly static anyway so I also increased the default timeout but you can adjust it to test the recovery

you can get the latest CI build here https://github.com/Velocidex/velociraptor/actions/runs/3234789607

Thanks for reporting this!

**Comment by randomaccess3** [comment]
Some VQL that might be useful to identify KAPE collections that have failures

`select * from foreach(
    row = {
        select client_id AS ClientId, os_info.hostname AS hostname from clients()
    },
    query = {
        select * from foreach(
        row = {select session_id from flows(client_id=ClientId) where artifacts_with_results =~ "Windows.KapeFiles.Targets"},
        query = {
            SELECT ClientId, hostname, session_id, "KAPE Error" As Description FROM flow_logs(client_id=ClientId, flow_id=session_id) Where message =~ "serviceprofile.*used.*process"
        }
        )
    }
)`

**Comment by scudette** [comment]
This should be fixed by #2157
